macros.tex

```diff
@@ -676,18 +676,77 @@
+\setcounter{@newthmdenom}{1}
+% This macro will be called to create and munge all theorem
+% environments, *including* the base one "thm" that we have already
+% created.
 \def\defthm#1#2#3{%
-  %% Ensure all theorem types are numbered with the same counter
-  \newaliascnt{#1}{thm}
-  \newtheorem{#1}[#1]{#2}
-  \aliascntresetthe{#1}
-  %% This command tells cleveref's \cref what to call things
-  \crefname{#1}{#2}{#3}}
+  % The actual "theorem environment" we use internally is not "thm" or
+  % "cor", but rather "my@thm" or "my@cor".  The aliascnt commands
+  % ensure that all such environments use the same counter "thm".
+  \newaliascnt{my@#1}{thm}
+  \newtheorem{my@#1}[my@#1]{#2}
+  \aliascntresetthe{my@#1}
+  % This command tells cleveref's \cref what to call things
+  \crefname{#1}{#2}{#3}
+  % Now we define "thm"/"cor"/etc. to be a wrapper around its my@
+  % version, which resets the "most recently seen fraction" counters
+  % afterwards.
+  \expandafter\def\csname #1\endcsname{\begin{my@#1}}
+  \expandafter\def\csname end#1\endcsname{\end{my@#1}\setcounter{@newthmnum}{0}\setcounter{@newthmdenom}{1}}
+  % Next we define the "newthm" environment for fractions.  Again, there
+  % is an internal "my@newthm" version.  Note that this does *not* use
+  % the "thm" counter, since we don't want that counter to get
+  % incremented when we pass through a newthm.  Instead it creates its
+  % own my@newthm counter, which we never actually use.
+  \newtheorem{my@new#1}{#2}
+  \crefname{my@new#1}{#2}{#3}
+  % Instead of using the "my@newthm" counter, when we need to number a
+  % newthm, we call the special macro \newthm defined below.
+  \expandafter\renewcommand\csname themy@new#1\endcsname{\thenewthm}
+  % Finally, we define the "newthm" environment to first check
+  % validity of the fraction passed and then create the theorem.
+  \newenvironment{new#1}[2]{\checknewthmfrac{##1}{##2}\setcounter{@newthmnum}{##1}\setcounter{@newthmdenom}{##2}\begin{my@new#1}}{\end{my@new#1}}
+}
+% How to display fractional theorem numbers.
+% (In particular, we leave off the zero if it is there)
+\def\thenewthm{\ifnum0<\value{my@thm}\themy@thm\else\thesection.\fi$\frac{\arabic{@newthmnum}}{\arabic{@newthmdenom}}$}
+% Reset the fraction at section boundaries too.  This is a bit of a
+% hack; \@seccntformat is not intended for this.
+\let\my@seccntformat\@seccntformat
+\renewcommand*{\@seccntformat}[1]{\setcounter{@newthmnum}{0}\setcounter{@newthmdenom}{1}\my@seccntformat{#1}}
+% Test to make sure newthm fractions increase and are all in between zero and one
+\newcount\@xone
+\newcount\@xtwo
+\def\checknewthmfrac#1#2{
+  \ifnum0<#1\else\errmessage{Fractional theorem numbers must be greater than zero}\fi
+  \ifnum#1<#2\else\errmessage{Fractional theorem numbers must be less than one}\fi
+  \@xone=#1
+  \@xtwo=#2
+  \multiply\@xtwo by \value{@newthmnum}
+  \multiply\@xone by \value{@newthmdenom}
+  \ifnum\@xone>\@xtwo\else\errmessage{Theorem numbers out of order}\fi
```

Comment:
I don't know if you ever plan on merging this, but if you do, it might be nice to give expect errors like
```suggestion
  \ifnum\@xone>\@xtwo\else\errmessage{Theorem numbers out of order: #1/#2 <= \arabic{@newthmnum}
  / \arabic{@newthmdenom}}\fi
```
(Or maybe we want `\the\value` instead of `\arabic`?)